### ISSUE TITLE
pwstore-fast-2.4.2... compilation error

### ISSUE THREAD

**Issue by agocorona** [PeterScott/pwstore#14]
```
    > cabal install pwstore-fast-2.4.2
   Configuring pwstore-fast-2.4.2...
   Building pwstore-fast-2.4.2...
   Failed to install pwstore-fast-2.4.2
   Last 10 lines of the build log ( /home/travis/.cabal/logs/pwstore-fast-2.4.2.log ):
   Crypto/PasswordStore.hs:334:95:
     Couldn't match expected type `[Char]' with actual type `ByteString'
     In the second argument of `constantTimeCompare', namely `goodHash'
     In the expression:
       encode (algorithm userInput salt (strengthModifier strength))
      `constantTimeCompare` goodHash
       In a case alternative:
          Just (strength, salt, goodHash)
            -> encode (algorithm userInput salt (strengthModifier strength))
               `constantTimeCompare` goodHash
```

**Comment by accursoft** [comment]
Getting this error with GHC 7.8.3.

**Comment by AxisOfEval** [comment]
Yeap. Can confirm this. cabal install blows up.

``` shell
Resolving dependencies...
Configuring pwstore-fast-2.4.2...
Building pwstore-fast-2.4.2...
Preprocessing library pwstore-fast-2.4.2...
[1 of 1] Compiling Crypto.PasswordStore ( Crypto/PasswordStore.hs, dist/build/Crypto/PasswordStore.o )

Crypto/PasswordStore.hs:334:11:
    Couldn't match expected type `[Char]' with actual type `ByteString'
    In the return type of a call of `encode'
    In the first argument of `constantTimeCompare', namely
      `encode (algorithm userInput salt (strengthModifier strength))'
    In the expression:
      encode (algorithm userInput salt (strengthModifier strength))
      `constantTimeCompare` goodHash

Crypto/PasswordStore.hs:334:95:
    Couldn't match expected type `[Char]' with actual type `ByteString'
    In the second argument of `constantTimeCompare', namely `goodHash'
    In the expression:
      encode (algorithm userInput salt (strengthModifier strength))
      `constantTimeCompare` goodHash
    In a case alternative:
        Just (strength, salt, goodHash)
          -> encode (algorithm userInput salt (strengthModifier strength))
             `constantTimeCompare` goodHash
Failed to install pwstore-fast-2.4.2
cabal: Error: some packages failed to install:
pwstore-fast-2.4.2 failed during the building phase. The exception was:
ExitFailure 1
snap-0.13.2.8 depends on pwstore-fast-2.4.2 which failed to install.
```

**Comment by mightybyte** [comment]
We're also seeing the error with GHC 7.6.3 as well.

**Comment by peti** [comment]
Here's one more build log illustrating the problem from NixOS: http://hydra.cryp.to/build/181211/nixlog/2/raw

**Comment by PeterScott** [comment]
Note to self: set up Travis CI to test this with multiple GHC versions. Apparently testing with just one isn't enough. :-(

I've upgraded my version of GHC, reproduced the bug, and switched to using the constant-time comparison function from byteable. This fixes it on my machine. Regrettably, I only have one version of GHC installed at a time, so who knows what it'll do on yours....

Anyway, I've pushed the fix for now, and will set up better CI.
